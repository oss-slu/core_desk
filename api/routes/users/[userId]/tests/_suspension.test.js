import { describe, expect, it } from "vitest";
import request from "supertest";
import { app } from "../../../../index.js";
import { gt } from "#gt";
import { prisma } from "#prisma";
import { prisma as mockPrisma } from "#mock-prisma";
import { LogType } from "@prisma/client";

describe("/[userId]", () => {
  describe("Suspend user", async () => {
    it("should return 403 if requesting user is not a global admin", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
        },
      });

      const res = await request(app)
        .post(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(403);
      expect(res.body).toEqual({ error: "Unauthorized" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
      });

      expect(updatedUser.suspended).toBe(false);

      const requestingUser = await prisma.user.findUnique({
        where: { email: "test@email.com" },
        include: {
          logs: {
            where: {
              type: LogType.FORBIDDEN_ACTION,
            },
          },
        },
      });

      expect(requestingUser.logs).toHaveLength(1);
      expect(requestingUser.logs[0].type).toBe(LogType.FORBIDDEN_ACTION);
    });

    it("should return 404 if user is not found", async () => {
      const res = await request(app)
        .post(`/api/users/_/suspension`)
        .set(...(await gt({ ga: true })))
        .send();

      expect(res.status).toBe(404);
      expect(res.body).toEqual({ error: "User not found" });

      expect(mockPrisma.logs.create).not.toHaveBeenCalled();
      expect(mockPrisma.user.update).not.toHaveBeenCalled();
    });

    it("should suspend user", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
        },
      });

      const res = await request(app)
        .post(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt({ ga: true })))
        .send({
          reason: "Test reason",
        });

      expect(res.status).toBe(200);
      expect(res.body).toEqual({ message: "User is now suspended" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
        include: {
          logs: {
            where: {
              type: LogType.USER_SUSPENSION_APPLIED,
            },
          },
        },
      });

      expect(updatedUser.suspended).toBe(true);
      expect(updatedUser.suspensionReason).toBe("Test reason");
      expect(updatedUser.logs).toHaveLength(1);
    });
  });

  describe("Unsuspend user", async () => {
    it("should return 403 if requesting user is not a global admin", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
          suspended: true,
        },
      });

      const res = await request(app)
        .delete(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt()))
        .send();

      expect(res.status).toBe(403);
      expect(res.body).toEqual({ error: "Unauthorized" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
      });

      expect(updatedUser.suspended).toBe(true);

      const requestingUser = await prisma.user.findUnique({
        where: { email: "test@email.com" },
        include: {
          logs: {
            where: {
              type: LogType.FORBIDDEN_ACTION,
            },
          },
        },
      });

      expect(requestingUser.logs).toHaveLength(1);
      expect(requestingUser.logs[0].type).toBe(LogType.FORBIDDEN_ACTION);
    });

    it("should return 404 if user is not found", async () => {
      const res = await request(app)
        .delete(`/api/users/_/suspension`)
        .set(...(await gt({ ga: true })))
        .send();

      expect(res.status).toBe(404);
      expect(res.body).toEqual({ error: "User not found" });

      expect(mockPrisma.logs.create).not.toHaveBeenCalled();
      expect(mockPrisma.user.update).not.toHaveBeenCalled();
    });

    it("should unsuspend user", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
          suspended: true,
        },
      });

      const res = await request(app)
        .delete(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt({ ga: true })))
        .send();

      expect(res.status).toBe(200);
      expect(res.body).toEqual({ message: "User is no longer suspended" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
        include: {
          logs: {
            where: {
              type: LogType.USER_SUSPENSION_REMOVED,
            },
          },
        },
      });

      expect(updatedUser.suspended).toBe(false);
      expect(updatedUser.logs).toHaveLength(1);
    });
  });

  describe("Change suspension reason", async () => {
    it("should return 403 if requesting user is not a global admin", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
          suspended: true,
          suspensionReason: "Test reason",
        },
      });

      const res = await request(app)
        .put(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt()))
        .send({
          reason: "New reason",
        });

      expect(res.status).toBe(403);
      expect(res.body).toEqual({ error: "Unauthorized" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
      });

      expect(updatedUser.suspended).toBe(true);
      expect(updatedUser.suspensionReason).toBe("Test reason");

      const requestingUser = await prisma.user.findUnique({
        where: { email: "test@email.com" },
        include: {
          logs: {
            where: {
              type: LogType.FORBIDDEN_ACTION,
            },
          },
        },
      });

      expect(requestingUser.logs).toHaveLength(1);
      expect(requestingUser.logs[0].type).toBe(LogType.FORBIDDEN_ACTION);
    });

    it("should return 403 if target user is not suspended", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
          suspended: false,
        },
      });

      const res = await request(app)
        .put(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt({ ga: true })))
        .send({
          reason: "New reason",
        });

      expect(res.status).toBe(400);
      expect(res.body).toEqual({ error: "User is not suspended" });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
        include: {
          logs: {
            where: {
              type: LogType.USER_SUSPENSION_CHANGED,
            },
          },
        },
      });

      expect(updatedUser.suspended).toBe(false);
      expect(updatedUser.suspensionReason).toBeNull();
      expect(updatedUser.logs).toHaveLength(0);
    });

    it("should return 404 if user is not found", async () => {
      const res = await request(app)
        .put(`/api/users/_/suspension`)
        .set(...(await gt({ ga: true })))
        .send({
          reason: "New reason",
        });

      expect(res.status).toBe(404);
      expect(res.body).toEqual({ error: "User not found" });

      expect(mockPrisma.logs.create).not.toHaveBeenCalled();
      expect(mockPrisma.user.update).not.toHaveBeenCalled();
    });

    it("should change suspension reason", async () => {
      const createdUser = await prisma.user.create({
        data: {
          email: "toBeSuspended@test.com",
          firstName: "John",
          lastName: "Doe",
          suspended: true,
          suspensionReason: "Test reason",
        },
      });

      const res = await request(app)
        .put(`/api/users/${createdUser.id}/suspension`)
        .set(...(await gt({ ga: true })))
        .send({
          reason: "New reason",
        });

      expect(res.status).toBe(200);
      expect(res.body).toEqual({
        message: "Suspension reason has been updated",
      });

      const updatedUser = await prisma.user.findUnique({
        where: { id: createdUser.id },
        include: {
          logs: {
            where: {
              type: LogType.USER_SUSPENSION_CHANGED,
            },
          },
        },
      });

      expect(updatedUser.suspended).toBe(true);
      expect(updatedUser.suspensionReason).toBe("New reason");
      expect(updatedUser.logs).toHaveLength(1);
    });
  });
});
