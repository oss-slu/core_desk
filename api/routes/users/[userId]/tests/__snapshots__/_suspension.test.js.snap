// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/[userId] > Change suspension reason > should return 403 if requesting user is not a global admin 1`] = `
{
  "billingGroupId": null,
  "billingGroupInvitationLinkId": null,
  "commentId": null,
  "createdAt": 2024-11-19T18:10:05.672Z,
  "from": null,
  "id": Any<String>,
  "jobId": null,
  "jobItemId": null,
  "ledgerItemId": null,
  "materialId": null,
  "materialImageId": null,
  "message": "{"method":"PUT","url":"/api/users/cm3ors3c400145fu8lvidrazb/suspension","body":{"reason":"New reason"},"query":{},"ip":"::ffff:127.0.0.1"}",
  "resourceId": null,
  "resourceImageId": null,
  "resourceTypeId": null,
  "shopId": null,
  "to": null,
  "type": "FORBIDDEN_ACTION",
  "updatedAt": 2024-11-19T18:10:05.672Z,
  "userBillingGroupId": null,
  "userId": Any<String>,
}
`;

exports[`/[userId] > Suspend user > should return 403 if requesting user is not a global admin 1`] = `
{
  "billingGroupId": null,
  "billingGroupInvitationLinkId": null,
  "commentId": null,
  "createdAt": 2024-11-19T18:10:05.497Z,
  "from": null,
  "id": Any<String>,
  "jobId": null,
  "jobItemId": null,
  "ledgerItemId": null,
  "materialId": null,
  "materialImageId": null,
  "message": "{"method":"POST","url":"/api/users/cm3ors36b00045fu89whnoc22/suspension","body":{},"query":{},"ip":"::ffff:127.0.0.1"}",
  "resourceId": null,
  "resourceImageId": null,
  "resourceTypeId": null,
  "shopId": null,
  "to": null,
  "type": "FORBIDDEN_ACTION",
  "updatedAt": 2024-11-19T18:10:05.497Z,
  "userBillingGroupId": null,
  "userId": Any<String>,
}
`;

exports[`/[userId] > Unsuspend user > should return 403 if requesting user is not a global admin 1`] = `
{
  "billingGroupId": null,
  "billingGroupInvitationLinkId": null,
  "commentId": null,
  "createdAt": 2024-11-19T18:10:05.596Z,
  "from": null,
  "id": Any<String>,
  "jobId": null,
  "jobItemId": null,
  "ledgerItemId": null,
  "materialId": null,
  "materialImageId": null,
  "message": "{"method":"DELETE","url":"/api/users/cm3ors39z000m5fu8edbetcyy/suspension","body":{},"query":{},"ip":"::ffff:127.0.0.1"}",
  "resourceId": null,
  "resourceImageId": null,
  "resourceTypeId": null,
  "shopId": null,
  "to": null,
  "type": "FORBIDDEN_ACTION",
  "updatedAt": 2024-11-19T18:10:05.596Z,
  "userBillingGroupId": null,
  "userId": Any<String>,
}
`;
